src/instrumentation/podmutator.go

```diff
@@ -35,150 +42,244 @@ var (
-	Php    *v1alpha1.Instrumentation
-	Ruby   *v1alpha1.Instrumentation
-	Go     *v1alpha1.Instrumentation
+	client                 client.Client
+	sdkInjector            SdkInjector
+	secretReplicator       SecretReplicator
+	instrumentationLocator InstrumentationLocator
+	logger                 logr.Logger
+	operatorNamespace      string
 }
 
-var _ webhookhandler.PodMutator = (*instPodMutator)(nil)
-
-func NewMutator(logger logr.Logger, client client.Client) *instPodMutator {
+// NewMutator is used to get a new instance of a mutator
+func NewMutator(
+	logger logr.Logger,
+	client client.Client,
+	sdkInjector SdkInjector,
+	secretReplicator SecretReplicator,
+	instrumentationLocator InstrumentationLocator,
+	operatorNamespace string,
+) *instPodMutator {
 	return &instPodMutator{
-		Logger: logger,
-		Client: client,
-		sdkInjector: &sdkInjector{
-			logger: logger,
-			client: client,
-		},
+		logger:                 logger,
+		client:                 client,
+		instrumentationLocator: instrumentationLocator,
+		sdkInjector:            sdkInjector,
+		secretReplicator:       secretReplicator,
+		operatorNamespace:      operatorNamespace,
 	}
 }
 
+// Mutate is used to mutate a pod based on some instrumentation(s)
 func (pm *instPodMutator) Mutate(ctx context.Context, ns corev1.Namespace, pod corev1.Pod) (corev1.Pod, error) {
-	logger := pm.Logger.WithValues("namespace", pod.Namespace, "name", pod.Name)
-
-	var inst *v1alpha1.Instrumentation
-	var err error
-
-	insts := languageInstrumentations{}
-
-	// We bail out if any annotation fails to process.
+	logger := pm.logger.WithValues("namespace", pod.Namespace, "name", pod.Name)
 
-	if inst, err = pm.getInstrumentationInstance(ctx, ns, pod, annotationInjectJava); err != nil {
+	instCandidates, err := pm.instrumentationLocator.GetInstrumentations(ctx, ns, pod)
+	if err != nil {
 		// we still allow the pod to be created, but we log a message to the operator's logs
 		logger.Error(err, "failed to select a New Relic Instrumentation instance for this pod")
 		return pod, err
 	}
-	insts.Java = inst
+	if len(instCandidates) == 0 {
+		logger.Info("no New Relic Instrumentation instance for this Pod")
+		return pod, errNoInstancesAvailable
+	}
 
-	if inst, err = pm.getInstrumentationInstance(ctx, ns, pod, annotationInjectNodeJS); err != nil {
-		// we still allow the pod to be created, but we log a message to the operator's logs
-		logger.Error(err, "failed to select a New Relic Instrumentation instance for this pod")
+	instrumentations, err := GetLanguageInstrumentations(instCandidates)
+	if err != nil {
+		if errors.Is(err, errMultipleInstancesPossible) {
+			pm.logger.Info("too many New Relic Instrumentation instances for this Pod.  only 1 allowed")
+		} else {
+			logger.Error(err, "failed to select a New Relic Instrumentation instance for this Pod")
+		}
 		return pod, err
 	}
-	insts.NodeJS = inst
+	mutatedPod := pm.sdkInjector.Inject(ctx, instrumentations, ns, pod)
 
```

Comment:
let's move this line to the last line before return otherwise we could be mutating the pod and then running into error later and not doing anything with the mutated pod.

Reply:
moved